elkm1_lib/connection.py

```diff
@@ -1,287 +1,188 @@
-
-    def _cancel_write_timer(self) -> None:
-        if self._write_timeout_task:
-            self._write_timeout_task.cancel()
-            self._write_timeout_task = None
-
-    def _cancel_heartbeat_timer(self) -> None:
-        if self._heartbeat_timeout_task is not None:
-            self._heartbeat_timeout_task.cancel()
-
-    def _heartbeat_timeout(self) -> None:
-        LOG.warning("ElkM1 connection heartbeat timed out, disconnecting")
-        if self._transport:
-            self._transport.close()
-
-    def _restart_heartbeat_timer(self) -> None:
-        self._cancel_heartbeat_timer()
-        if self._heartbeat_time > 0:
-            self._heartbeat_timeout_task = self._loop.call_later(
-                self._heartbeat_time, self._heartbeat_timeout
-            )
-
-    def data_received(self, data: bytes) -> None:
-        self._restart_heartbeat_timer()
-        self._buffer += data.decode("ISO-8859-1")
-        while "\r\n" in self._buffer:
-            line, self._buffer = self._buffer.split("\r\n", 1)
-            if get_elk_command(line) == self._waiting_for_response:
-                self._waiting_for_response = None
-                self._cancel_write_timer()
-            self._got_data_callback(line)
-        self._process_write_queue()
-
-    def _process_write_queue(self) -> None:
-        while self._queued_writes and not self._waiting_for_response:
-            to_write = self._queued_writes.pop(0)
-            self.write_data(to_write[0], to_write[1], timeout=to_write[2])
-
-    def write_data(
-        self,
-        data: str,
-        response_required: str | None = None,
-        timeout: float = 5.0,
-        raw: bool = False,
-    ) -> None:
-        """Write data on the asyncio Protocol"""
-        if self._transport is None:
-            return
-
-        if self._paused:
-            return
-
-        if self._waiting_for_response:
-            LOG.debug("queueing write %s", data)
-            self._queued_writes.append((data, response_required, timeout))
-            return
-
-        if response_required:
-            self._waiting_for_response = response_required
-            if timeout > 0:
-                self._write_timeout_task = self._loop.call_later(
-                    timeout, self._response_required_timeout
-                )
-
-        if not raw:
-            cksum = (256 - reduce(lambda x, y: x + y, map(ord, data))) % 256
-            data = f"{data}{cksum:02X}"
-            if int(data[0:2], 16) != len(data) - 2:
-                LOG.debug("message length wrong: %s", data)
+    def disconnect(self, reason: str = "") -> None:
+        """Disconnect and cleanup."""
+        LOG.warning("ElkM1 disconnecting %s", reason)
+        if self._writer:
+            self._writer.close()
+            self._writer = None
+        self._write_queue.clear()
+        for task in self._tasks:
+            if asyncio.current_task() != task:
+                task.cancel()
```

Comment:
Nothing is retrieving any potential future exceptions in these tasks after cancelation because they aren't being awaited after cancelation.

Reply:
I admit to being a novice here...

In all my reading there is no requirement to catch the `CancelledError`, it will just be ignored in the task if not caught when the task is shutting down. The `CancelledError` can be caught if there is some cleanup to do - but I can't think of any cleanup required in any of the three coroutines.

I put two bits of test code in to check assumptions. The first bit is `loop.set_debug(True)`. Nothing interesting is printed. If I use `readuntil` (as suggested in your previous comment) then I get an unprocessed exception.

The second bit of added test code is in the `connection` method I added `add_done_callback` for `all_tasks`. Again, nothing interesting is printed from that, just the completions as would be expected. Both of these changes will are pushed as temporary code.

Based on all the testing and reading I can't see what I would change in this bit of code. Also, `readuntil` doesn't not really simplify this Elk reading code, mostly because then I would have to handle the exceptions.

I don't know if I have missed something in the many test scenarios and bits of test code. Do you see something?